core/src/main/scala/org/apache/spark/util/JsonProtocol.scala

```diff
@@ -439,7 +439,14 @@ private[spark] object JsonProtocol {
     val reason = Utils.getFormattedClassName(taskEndReason)
     val json: JObject = taskEndReason match {
       case fetchFailed: FetchFailed =>
-        val blockManagerAddress = Option(fetchFailed.bmAddress).
+        val blockManagerAddress = Option {
+          fetchFailed.bmAddress match {
+            case bmId: BlockManagerId =>
+              bmId
+            case _ =>
+              // TODO: support custom locations
+              null
+          }}.
           map(blockManagerIdToJson).getOrElse(JNothing)
         ("Block Manager Address" -> blockManagerAddress) ~
```

Comment:
I'm think of replacing `("Block Manager Address" -> blockManagerAddress)` with ("custom location name" -> locationString) when the `bmAddress` is a custom location. Thus, we'd be able to log custom location as well as keep the backward compatibility unless users log the custom location but read it uses old version.